### DISCUSSION TITLE
Various M5Stack I2C modules with STM32 IO expansions

### DISCUSSION THREAD

**Discussion by lamium2** [arendst/Tasmota#21924] [Support request] [answered]
Hello everyone,
I would have liked to use the M5Stack system to build a controller for our garden shed, for which I would like to use the Programmable Power Supply Module to charge the robotic lawnmower and power the ESP32 (CoreS3SE controller).
M5Stack offers many more modules that connect an STM32 via I2C and provide numerous GPIO, I2C or other features behind it.

for example:
- M5Stack Ext Encoder Unit
- Programmable Power Supply Module 13.2
- 4-Relay Unit
- H-bridge Unit v1.1 (STM32F030)
- QR Code Scanner Unit (STM32F030)
- 8-Channel Servo Driver Unit (STM32F030)
- BLDC Motor Driver Unit (STM32)
- 8-Encoder Unit (STM32F030)
- Extend I/O 2 Unit (STM32F0)
-  I/O Hub 1 to 6 Expansion Unit (STM32F0)
- Analog to I2C Unit 4-20mA Input (STM32G030)
-  4-Channel Encoder Motor Driver Module (STM32F030) v1.1
- Goplus2 DC Motor and Servo Driver Module (STM32F0)
- 2-Channel AC Relay Module 13.2 (STM32F030)
- 4IN8OUT Multi-channel DC Drive Module (STM32F030)
- 4-Channel Relay 13.2 Module V1.1 (STM32F030)
- 4-Channel Encoder Motor Driver Module (STM32F030)
- 4-Channel Analog to 12C 13.2 Module 4-20mA Input(ST32G030)

Is there an easy way (e.g. berry) to access it?
Is there a driver similar to the MCU23xx that simply passes the pins through?

The commands for the PPS are quite simple (see https://docs.m5stack.com/en/module/Module13.2-PPS?ref=8oPoadLv), but unfortunately I don't know how to send them via the STM32 on I2C 0x35.

Thanks in advance!

**Comment by lamium2** [comment]
M5Stack provides the following code, is it usable?
https://github.com/m5stack/M5Module-PPS/tree/main/src

**Comment by arendst** [comment, marked as answer]
Until now you have to write the I2C drivers yourselves. Either in C++ or Berry. Except for the libraries (which are just simple I2C calls already present in Tasmota) you'll need code to access the specific functionalities. Like the 4 relay module, you'll have to write code to access any relay from Tasmota.

Currently tasmota supports ESP8266/ESP32 GPIO's where you can easily connect relays, motors, lights. The M5Stack can do the same when I2C drivers are made in addition to the M5 provided libraries.
